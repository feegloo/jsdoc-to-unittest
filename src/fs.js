import fs from 'fs';
import util from 'util';

export const writeFileAsync = util.promisify(fs.writeFile);
export const readFileAsync = util.promisify(fs.readFile);

export const toStdout = (output) => {
  if (process.env.NODE_ENV === 'test') {
    return output;
  }

  process.stdout.write(output);
  return true;
};
