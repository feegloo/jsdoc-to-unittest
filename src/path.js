export default ([base, ...parts], isFunction) => {
  let path = base;
  parts.forEach((part) => {
    if (Number.isNaN(+part) || typeof part !== 'string') {
      path += `.${part}`;
    } else {
      path += `[${part}]`;
    }
  });
  if (isFunction) {
    path += '()';
  }

  return path;
};
