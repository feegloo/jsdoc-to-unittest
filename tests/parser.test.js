import fs from 'fs';
import util from 'util';
import path from 'path';
import parse from 'src/parser';

const readFileAsync = util.promisify(fs.readFile);

describe('Parser', () => {
  test('happy path works', async () => {
    const { exports, imports, samples } = await parse(
      await readFileAsync(path.resolve(__dirname, 'fixtures/contains.js'), 'utf-8'),
    );

    expect(exports).toHaveLength(1);
    expect(imports).toHaveLength(1);
    expect(Function(`${exports[0]};return ${imports[0]}`)()).toBeInstanceOf(Function);
    expect(samples).toHaveLength(1);
    expect(samples[0].examples).toHaveLength(2);
    expect(samples[0].name).toBe('contains');
    expect(samples[0].examples).toMatchSnapshot();
  });

  test('tries resolving the name', async () => {
    const { samples } = await parse(
      await readFileAsync(path.resolve(__dirname, 'fixtures/contains-no-name.js'), 'utf-8'),
    );

    expect(samples[0].name).toBe('contains');
  });

  test('multiline examples', async () => {
    const { exports, imports, samples } = await parse(
      await readFileAsync(path.resolve(__dirname, 'fixtures/each.js'), 'utf-8'),
    );

    expect(exports).toHaveLength(1);
    expect(imports).toHaveLength(1);
    expect(samples).toHaveLength(1);
    expect(samples[0].name).toBe('each');
    expect(samples[0].examples).toMatchSnapshot();
  });

  test('grabs @returns', async () => {
    const { samples } = await parse(`/**
     * @name contains
     *
     * @example
     * contains('hey', 'ey');
     *
     * @param str1
     * @param str2
     * @returns {*|boolean}
     */
    function contains(str1, str2) {
      return str1 && str1.indexOf(str2) !== -1;
    }`);

    expect(samples[0].examples).toEqual([
      {
        code: "contains('hey', 'ey');", name: 'contains', result: ['*', 'boolean'], type: 'instance',
      }],
    );
  });

  test('detects mocks', async () => {
    const { samples } = await parse(`/**
     * @name contains
     *
     * @example
     * contains('hey', 'ey'); // true
     * @mock { 'contains(1,2)': false }
     *
     * @param str1
     * @param str2
     * @returns {*|boolean}
     */
    function contains(str1, str2) {
      return str1 && str1.indexOf(str2) !== -1;
    }`);

    expect(samples[0].examples).toMatchSnapshot();
  });

  test('supports @async', async () => {
    expect(await parse(
      await readFileAsync(path.resolve(__dirname, 'fixtures/async.js'), 'utf-8'),
    )).toMatchSnapshot();
  });

  test('detects async automatically', async () => {
    expect(await parse(
      await readFileAsync(path.resolve(__dirname, 'fixtures/get-device-price.js'), 'utf-8'),
    )).toMatchSnapshot();
  });

  test('numeral-conjugation.js is parsed correctly', async () => {
    const { exports, imports, samples } = await parse(
      await readFileAsync(path.resolve(__dirname, 'fixtures/numeral-conjugation.js'), 'utf-8'),
    );

    expect(exports).toHaveLength(1);
    expect(imports).toHaveLength(1);
    expect(samples).toHaveLength(1);
    expect(samples[0].examples).toMatchSnapshot();
  });
});
