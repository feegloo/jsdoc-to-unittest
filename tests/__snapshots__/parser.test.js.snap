// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser detects async automatically 1`] = `
Object {
  "exports": Array [
    "function getDevicePrice() {
    return easy.domElements.waitFor({ selector: '.device-price' }).then(function (els) {
        return site.utils.extractPrice(els[0].textContent);
    });
}",
  ],
  "imports": Array [
    "getDevicePrice",
  ],
  "samples": Array [
    Object {
      "examples": Array [
        Object {
          "async": 1,
          "code": "location.href = 'http://www.play.pl/telefony/glosniki/glosnik-jabra-speak-410/?oid=4000963609';

frosmo.site.products.getDevicePrice()
.then(function(devicePrice){
     console.log(devicePrice);  // 123
});",
          "name": undefined,
          "result": "123",
          "type": "value",
        },
      ],
      "name": undefined,
      "tags": Object {
        "async": Array [],
        "example": Array [
          Object {
            "description": "location.href = 'http://www.play.pl/telefony/glosniki/glosnik-jabra-speak-410/?oid=4000963609';

frosmo.site.products.getDevicePrice()
.then(function(devicePrice){
     console.log(devicePrice);  // 123
});",
            "title": "example",
          },
        ],
        "memberof": Array [
          Object {
            "description": "site.products",
            "title": "memberof",
          },
        ],
        "mock": Array [],
        "name": Array [
          Object {},
        ],
        "param": Array [
          Object {
            "description": "window",
            "name": "global",
            "title": "param",
            "type": Object {
              "expression": Object {
                "name": "Window",
                "type": "NameExpression",
              },
              "type": "OptionalType",
            },
          },
        ],
        "returns": Array [
          Object {
            "description": null,
            "title": "return",
            "type": Object {
              "name": "easy.Promise",
              "type": "NameExpression",
            },
          },
        ],
      },
    },
  ],
}
`;

exports[`Parser detects mocks 1`] = `
Array [
  Object {
    "code": "contains('hey', 'ey'); // true",
    "mocks": Object {
      "contains(1,2)": false,
    },
    "name": "contains",
    "result": "true",
    "type": "value",
  },
]
`;

exports[`Parser happy path works 1`] = `
Array [
  Object {
    "code": "contains('foo', 'fooo'); // false",
    "name": "contains",
    "result": "false",
    "type": "value",
  },
  Object {
    "code": "contains('hey', 'ey'); // true;",
    "name": "contains",
    "result": "true",
    "type": "value",
  },
]
`;

exports[`Parser multiline examples 1`] = `
Array [
  Object {
    "code": "easy.utils.each([1, 2, 3], function (value, index) {
    easy.console.log(index, value);
});",
    "name": "each",
    "result": "",
    "type": "no-throw",
  },
  Object {
    "code": "easy.utils.each({foo: 'bar', baz: 'bar'}, function (value, key) {
    easy.console.log(key, value);
});",
    "name": "each",
    "result": "",
    "type": "no-throw",
  },
  Object {
    "code": "var utils = {
    power: function (base, exponent) {
        return Math.pow(base, exponent);
    }
};

easy.utils.each([1, 2, 3], function (value, index) {
    // \\"this\\" references the utils object where the function \\"power\\" is available
    easy.console.log(index, value, this.power(value));
}, utils);",
    "name": "each",
    "result": "",
    "type": "no-throw",
  },
  Object {
    "code": "easy.utils.each({foo: 'bar', baz: 'bar'}, function (value, key) {
    // \\"this\\" references the global object where the built-in String object is available
    easy.console.log(key, value, this.String.prototype.toUpperCase.call(value));
});",
    "name": "each",
    "result": "",
    "type": "no-throw",
  },
]
`;

exports[`Parser numeral-conjugation.js is parsed correctly 1`] = `
Array [
  Object {
    "code": "numeralConjugation(3, ['os贸b', 'osoba', 'osoby']); // '3 osoby'",
    "name": "numeralConjugation",
    "result": "\\"3 osoby\\"",
    "type": "value",
  },
  Object {
    "code": "numeralConjugation(11, ['os贸b', 'osoba', 'osoby']); // '11 os贸b'",
    "name": "numeralConjugation",
    "result": "\\"11 os贸b\\"",
    "type": "value",
  },
]
`;

exports[`Parser supports @async 1`] = `
Object {
  "exports": Array [
    "function getFromURL(str) {
    return fetch(str);
}",
  ],
  "imports": Array [
    "getFromURL",
  ],
  "samples": Array [
    Object {
      "examples": Array [
        Object {
          "async": 1,
          "code": "getFromURL('https://mail.google.com/mail/')
  .then(response => response.slice(5).trim()); // \\"world\\"",
          "mocks": Object {
            "fetch('string')": "hello world  ",
          },
          "name": "getFromURL",
          "result": "\\"world\\"",
          "type": "value",
        },
      ],
      "name": "getFromURL",
      "tags": Object {
        "async": Array [
          Object {
            "description": null,
            "title": "async",
          },
        ],
        "example": Array [
          Object {
            "description": "getFromURL('https://mail.google.com/mail/')
  .then(response => response.slice(5).trim()); // \\"world\\"",
            "title": "example",
          },
        ],
        "mock": Array [
          Object {
            "description": "{ 'fetch(\\\\'string\\\\')': 'hello world  ' }",
            "title": "mock",
          },
        ],
        "name": Array [
          Object {
            "description": null,
            "name": "getFromURL",
            "title": "name",
          },
        ],
        "param": Array [
          Object {
            "description": null,
            "name": "str1",
            "title": "param",
            "type": null,
          },
          Object {
            "description": null,
            "name": "str2",
            "title": "param",
            "type": null,
          },
        ],
        "returns": Array [
          Object {
            "description": null,
            "title": "returns",
            "type": Object {
              "elements": Array [
                Object {
                  "type": "AllLiteral",
                },
                Object {
                  "name": "boolean",
                  "type": "NameExpression",
                },
              ],
              "type": "UnionType",
            },
          },
        ],
      },
    },
  ],
}
`;
